rust/worker/src/execution/operators/brute_force_knn.rs

```diff
@@ -17,45 +21,103 @@ pub struct BruteForceKnnOperator {}
 /// * `distance_metric` - The distance metric to use.
 #[derive(Debug)]
 pub struct BruteForceKnnOperatorInput {
-    pub data: Vec<Vec<f32>>,
+    pub data: DataChunk,
     pub query: Vec<f32>,
     pub k: usize,
     pub distance_metric: DistanceFunction,
 }
 
 /// The output of the brute force k-nearest neighbors operator.
 /// # Parameters
+/// * `data` - The vectors to query against. This is a mask against the `data` input.
 /// * `indices` - The indices of the nearest neighbors. This is a mask against the `query_vecs` input.
 /// One row for each query vector.
 /// * `distances` - The distances of the nearest neighbors.
 /// One row for each query vector.
 #[derive(Debug)]
 pub struct BruteForceKnnOperatorOutput {
+    pub data: DataChunk,
     pub indices: Vec<usize>,
     pub distances: Vec<f32>,
 }
 
 pub type BruteForceKnnOperatorResult = Result<BruteForceKnnOperatorOutput, ()>;
 
+#[derive(Debug)]
+struct Entry {
+    index: usize,
+    distance: f32,
+}
+
+impl Ord for Entry {
+    fn cmp(&self, other: &Self) -> Ordering {
+        if self.distance == other.distance {
+            Ordering::Equal
+        } else if self.distance > other.distance {
+            // This is a min heap, so we need to reverse the ordering.
+            Ordering::Less
+        } else {
+            // This is a min heap, so we need to reverse the ordering.
+            Ordering::Greater
+        }
+    }
+}
+
+impl PartialOrd for Entry {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for Entry {
+    fn eq(&self, other: &Self) -> bool {
+        self.distance == other.distance
+    }
+}
+
+impl Eq for Entry {}
+
 #[async_trait]
 impl Operator<BruteForceKnnOperatorInput, BruteForceKnnOperatorOutput> for BruteForceKnnOperator {
     type Error = ();
 
     async fn run(&self, input: &BruteForceKnnOperatorInput) -> BruteForceKnnOperatorResult {
-        // We could use a heap approach here, but for now we just sort the distances and take the
-        // first k.
-        let mut sorted_indices_distances = input
-            .data
-            .iter()
-            .map(|data| input.distance_metric.distance(&input.query, data))
-            .enumerate()
-            .collect::<Vec<(usize, f32)>>();
-        sorted_indices_distances.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap());
-        let (sorted_indices, sorted_distances) = sorted_indices_distances
-            .drain(..cmp::min(input.k, input.data.len()))
-            .unzip();
+        let mut heap = BinaryHeap::with_capacity(input.k);
```

Comment:
for future (thinking out loud): moving this to binary heap would have been better as separate PR